### PR TITLE
MM-17276: Migrate 'components/toast_wrapper' and associated tests to TypeScript

### PR DESCRIPTION
**Pull request by diegovega49**
Summary:

Convert components/toast_wrapper and associated tests to typescript as part of the MatterMost typescript migration campaign.

Ticket link:

https://github.com/mattermost/mattermost-server/issues/17276
https://mattermost.atlassian.net/browse/MM-34347

Release note:

```release-note
NONE
```

Co-authored by: Diego Vega <[diegovega2249@gmail.com](mailto:diegovega2249@gmail.com)>
Co-authored by: Kent Nguyen <[nguyen.k.kent@gmail.com](mailto:nguyen.k.kent@gmail.com)>
Co-authored by: Michelle Tan <[michelle0223@gmail.com](mailto:michelle0223@gmail.com)>

### CODE REVIEW

**Comment by sinansonmez** [comment]
@M-ZubairAhmed  In order to support you, I can give a first look to this PR as reviewer (as I have a bit of experience migrating TypeScript files). If it is OK for you, please let me know and add me as reviewer (if possible, as I don't have the merge access to this repo)

File: components/toast_wrapper/toast_wrapper.tsx [hunk 1]
```diff
@@ -2,103 +2,120 @@
...
-             * Action creator to update toast status
-             */
-            updateToastStatus: PropTypes.func.isRequired,
-        }).isRequired,
+type OwnProps = {
+    unreadCountInChannel: number;
+    newRecentMessagesCount: number;
+    channelMarkedAsUnread?: boolean;
+    isCollapsedThreadsEnabled?: boolean;
+    rootPosts?: IDMappedObjects<Post>;
```
Comment (M-ZubairAhmed) [inline components/toast_wrapper/toast_wrapper.tsx:40]: Checkout how we are using PropsFromRedux in `components/emoji_picker/emoji_picker.tsx`, that way we dont have to rewrite types from redux again.

File: components/toast_wrapper/index.tsx [hunk 2]
```diff
@@ -2,37 +2,60 @@
...
+import {bindActionCreators, Dispatch} from 'redux';
+import {withRouter, RouteComponentProps} from 'react-router-dom';
+
+import {IDMappedObjects} from '@mattermost/types/utilities';
+
+import {Post} from 'packages/types/src/posts';
+
+import {GlobalState as ToastWrapperState} from '@mattermost/types/store';
+
+import {ViewsState} from 'types/store/views';
```
Comment (M-ZubairAhmed) [inline components/toast_wrapper/index.tsx:14]: nit picking, please rearrange the props how we are doing in `components/emoji_picker/emoji_picker.tsx` for consistency

File: components/toast_wrapper/index.tsx [hunk 3]
```diff
@@ -2,37 +2,60 @@
...
-            return Object.values(allPosts).filter((post) => {
-                return (
-                    post.root_id === '' &&
-                    post.channel_id === channel.id &&
-                    post.state !== Posts.POST_DELETED
-                );
+            return Object.values(allPosts).filter((post: Post) => {
+                return (post.root_id === '' &&
+                post.channel_id === channel.id &&
+                post.state !== Posts.POST_DELETED);
```
Comment (M-ZubairAhmed) [inline components/toast_wrapper/index.tsx:56]: nit: indenting
```suggestion
                return (
                    post.root_id === '' &&
                    post.channel_id === channel.id &&
                    post.state !== Posts.POST_DELETED
                 );
```

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Comment by M-ZubairAhmed** [comment]
> @M-ZubairAhmed @hmhealey In order to support you, I can give a first look to this PR as reviewer (as I have a bit of experience migrating TypeScript files). If it is OK for you, please let me know and add me as reviewer (if possible, as I don't have the merge access to this repo)

Please go ahead and leave your review, because it seems that for me to add you as reviewer you would be needed to add to MM organisation. 
image

But this still shouldnt stop you from leaving comments on a PR. can you check that

File: components/toast_wrapper/index.tsx [hunk 4]
```diff
@@ -2,37 +2,60 @@
...
             }).reduce((map, obj) => {
```
Comment (sinansonmez) [inline components/toast_wrapper/index.tsx:57]: type of the map should be `Record<string, boolean>` 

```suggestion
            }).reduce((map: Record<string, boolean>, obj) => {
```

Comment (sinansonmez) [inline components/toast_wrapper/index.tsx:57]: type of `map` should be `Record<string, boolean>`. Type of `rootPosts` should be adjusted accordingly in `toast_wrapper.tsx`

```suggestion
            }).reduce((map: Record<string, boolean>, obj) => {
```

File: components/toast_wrapper/index.tsx [hunk 5]
```diff
@@ -2,37 +2,60 @@
...
-                map[obj.id] = true;
+                (map as any)[obj.id] = true;
```
Comment (sinansonmez) [inline components/toast_wrapper/index.tsx:58]:
```suggestion
                map[obj.id] = true;
```

File: components/toast_wrapper/toast_wrapper.tsx [hunk 6]
```diff
@@ -2,103 +2,120 @@
...
+    constructor(props: Props) {
         super(props);
+        this.mounted = true;
         this.state = {
             unreadCountInChannel: props.unreadCountInChannel,
         };
     }
 
-    static countNewMessages = (postListIds, rootPosts, isCollapsedThreadsEnabled) => {
+    static countNewMessages = (postListIds: string[], rootPosts: IDMappedObjects<Post>, isCollapsedThreadsEnabled: boolean) => {
```
Comment (sinansonmez) [inline components/toast_wrapper/toast_wrapper.tsx:98]: This should eliminate the need for castings on line 118. 
```suggestion
    static countNewMessages = (postListIds: string[], rootPosts: Record<string, boolean>, isCollapsedThreadsEnabled?: boolean) => {
```

File: components/toast_wrapper/toast_wrapper.tsx [hunk 7]
```diff
@@ -2,103 +2,120 @@
...
+    static getDerivedStateFromProps(props: Props, prevState: State) {
         let {showUnreadToast, showNewMessagesToast, showMessageHistoryToast, showUnreadWithBottomStartToast} = prevState;
         let unreadCount;
 
         if (props.atLatestPost) {
             if (props.unreadScrollPosition === Preferences.UNREAD_SCROLL_POSITION_START_FROM_NEWEST && prevState.unreadCountInChannel) {
                 unreadCount = prevState.unreadCountInChannel + props.newRecentMessagesCount;
             } else {
-                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts, props.isCollapsedThreadsEnabled);
+                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts as IDMappedObjects<Post>, props.isCollapsedThreadsEnabled as boolean);
```
Comment (sinansonmez) [inline components/toast_wrapper/toast_wrapper.tsx:118]: As per my previous comment, if you change `rootPosts` to `Record<string, boolean>` and make it required (i.e. remove `?`) in Props type, you don't need to cast as `IDMappedObjects<Post>`

For `props.isCollapsedThreadsEnabled`, you just need to make it optional with `?`. Please see my suggestion above for `countNewMessages` definition

```suggestion
                unreadCount = ToastWrapper.countNewMessages(props.postListIds, props.rootPosts, props.isCollapsedThreadsEnabled);
```

File: components/toast_wrapper/toast_wrapper.tsx [hunk 8]
```diff
@@ -383,10 +400,10 @@ class ToastWrapper extends React.PureComponent {
             showActions: !atLatestPost || (atLatestPost && !atBottom),
         };
 
-        if (showUnreadToast && unreadCount > 0) {
+        if (showUnreadToast && unreadCount as number > 0) {
```
Comment (sinansonmez) [inline components/toast_wrapper/toast_wrapper.tsx:403]: If you check `unreadCount` before, you don't need to cast as number, and it is safer. 
```suggestion
        if (showUnreadToast && unreadCount && unreadCount > 0) {
```

File: components/toast_wrapper/toast_wrapper.tsx [hunk 9]
```diff
@@ -383,10 +400,10 @@ class ToastWrapper extends React.PureComponent {
...
             return (
                 <Toast {...unreadToastProps}>
-                    {this.newMessagesToastText(unreadCount, lastViewedAt)}
+                    {this.newMessagesToastText(unreadCount as number, lastViewedAt)}
```
Comment (sinansonmez) [inline components/toast_wrapper/toast_wrapper.tsx:406]:
```suggestion
                    {this.newMessagesToastText(unreadCount, lastViewedAt)}
```

File: components/toast_wrapper/toast_wrapper.tsx [hunk 10]
```diff
@@ -398,13 +415,13 @@ class ToastWrapper extends React.PureComponent {
...
-            jumpDirection: 'up',
+            jumpDirection: 'up' as const,
         };
 
-        if (showUnreadWithBottomStartToast && unreadCount > 0) {
+        if (showUnreadWithBottomStartToast && unreadCount as number > 0) {
             return (
                 <Toast {...unreadWithBottomStartToastProps}>
-                    {this.newMessagesToastText(unreadCount, lastViewedAt)}
+                    {this.newMessagesToastText(unreadCount as number, lastViewedAt)}
```
Comment (sinansonmez) [inline components/toast_wrapper/toast_wrapper.tsx:424]: Same comment here

**Review by sinansonmez — CHANGES_REQUESTED** [review]
@M-ZubairAhmed I was able to leave comments, thanks. 

@diegovega49 good job overall, I just have small comments about selected type casting for some of the props.

**Comment by M-ZubairAhmed** [comment]
@diegovega49 gentle ping on this one

File: components/toast_wrapper/index.tsx (hunk 3 above)
Comment (diegovega49) [reply components/toast_wrapper/index.tsx]: Updated

File: components/toast_wrapper/toast_wrapper.tsx (hunk 1 above)
Comment (diegovega49) [reply components/toast_wrapper/toast_wrapper.tsx:40]: So if my understanding is correct: with PropsfromRedux we will be exporting a type ConnectedProps which will contain prop type definitions? In which case I will modify ownProps to extend this type.

Comment (M-ZubairAhmed) [reply components/toast_wrapper/toast_wrapper.tsx:40]: yes exactly, we dont have to type props arising from redux explicitly.

**Comment by M-ZubairAhmed** [comment]
@diegovega49 lets get this to finish line

**Comment by lieut-data** [comment]
Heads up that as part of our efforts to move to a monorepo, we're closing out a number of older pull requests like this one to help streamline the effort.

If you'd like to preserve these changes -- even if you're not the original author! -- feel free to resubmit the pull request against the monorepo once it's ready. You can subscribe to [mattermost-server-issue-22420](https://www.github.com/mattermost/mattermost-server/issues/22420) for status updates on this effort.

**Pull request closed without merging**